File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-3.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    maxPathSum(root) {
        let res = root.val;

        function dfs(node){
            if(!node) return 0;

            let leftMax = Math.max(dfs(node.left), 0);
            let rightMax = Math.max(dfs(node.right), 0);

            res = Math.max(res, node.val + leftMax + rightMax);

            return node.val + Math.max(leftMax, rightMax);
        }

        dfs(root);
        return res;
    }   
}

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        def dfs(node):
            if not node:
                return
            if p.val == node.val:
                return p
            if q.val == node.val:
                return q

            if q.val > node.val and p.val > node.val:
                return dfs(node.right)
            if q.val < node.val and p.val < node.val:
                return dfs(node.left)
            
            return node
        
        return dfs(root)



